Autopilot/Src/imu.cpp

```diff
@@ -0,0 +1,219 @@
+    int16_t gyry = *(&(intImuDataPtr[5]));
+    int16_t gyrz = *(&(intImuDataPtr[6]));
+    int16_t accx = *(&(intImuDataPtr[7]));
+    int16_t accy = *(&(intImuDataPtr[8]));
+    int16_t accz = *(&(intImuDataPtr[9]));
+
+    Data.magx = static_cast<float> (magx);
+    Data.magy = static_cast<float> (magy);
+    Data.magz = static_cast<float> (magz);
+    Data.accx = static_cast<float> (accx) / ACC_RANGE_8_FACTOR;
+    Data.accy = static_cast<float> (accy) / ACC_RANGE_8_FACTOR;
+    Data.accz = static_cast<float> (accz) / ACC_RANGE_8_FACTOR;
+    Data.gyrx = static_cast<float> (gyrx) / GYRO_RANGE_1000_FACTOR;
+    Data.gyry = static_cast<float> (gyry) / GYRO_RANGE_1000_FACTOR;
+    Data.gyrz = static_cast<float> (gyrz) / GYRO_RANGE_1000_FACTOR;
+
+    Data.isDataNew = true;
+    Data.sensorStatus = 0;
+
+    BMX160::dataIsNew = false;
+}
+
+/***********************************************************************************************************************
+ * Private methods
+ **********************************************************************************************************************/
+
+BMX160::BMX160()
+{
+    HAL_GPIO_WritePin(GPIOG, GPIO_PIN_10, GPIO_PIN_SET);
+
+    SetAllPowerModesToNormal();
+    ConfigAcc();
+    ConfigGyro();
+    ConfigMag();
+
+    BMX160::dataIsNew = false;
+}
+
+void BMX160::SetAllPowerModesToNormal()
+{
+
+    Bmx160WriteReg(CMD_REG, ACC_NORMAL_MODE_CMD);
+    Bmx160WriteReg(CMD_REG, GYRO_NORMAL_MODE_CMD);
+    Bmx160WriteReg(CMD_REG, MAG_NORMAL_MODE_CMD);
+}
+
+void BMX160::ConfigAcc()
+{
+
+    Bmx160WriteReg(ACC_RANGE_REG, ACC_RANGE_8G);
+    Bmx160WriteReg(ACC_CONF_REG, ACC_ODR_800_OSR4);}
+
+
+void BMX160::ConfigGyro()
+{
+
+    Bmx160WriteReg(GYR_RANGE_REG, GYRO_RANGE_1000);
+    Bmx160WriteReg(GYR_CONF_REG, GYRO_ODR_800_OSR4);
+}
+
+void BMX160::ConfigMag()
+{
+    Bmx160WriteReg(MAG_IF_0_REG, MAG_SETUP_EN);
+
+
+    SetMagConfig(MAG_MODE_REG, MAG_SLEEP_MODE);
+    SetMagConfig(MAG_REPXY_REG, REP_XY_REGULAR_PRESET);
+    SetMagConfig(MAG_REPZ_REG, REP_Z_REGULAR_PRESET);
+
+    PrepareMagForDataMode();
+
+    Bmx160WriteReg(MAG_CONF_REG, MAG_REFRESH_50_HZ);
+    Bmx160WriteReg(MAG_IF_0_REG, MAG_SETUP_DIS);
+
+}
+
+void BMX160::SetMagConfig(uint8_t regAddr, uint8_t data)
+{
+    Bmx160WriteReg(MAG_IF_3_REG, data);
+    Bmx160WriteReg(MAG_IF_2_REG, regAddr);
+}
+
+void BMX160::PrepareMagForDataMode(void)
+{
+
+    Bmx160WriteReg(MAG_IF_3_REG, MAG_IF_3_DATA_MODE);
+    Bmx160WriteReg(MAG_IF_2_REG, MAG_IF_2_DATA_MODE);
+    Bmx160WriteReg(MAG_IF_1_REG, MAG_IF_1_DATA_MODE);
+}
+
+void BMX160::Bmx160WriteReg(uint8_t reg, uint8_t val)
+{
+    uint8_t rx[2] = { 0, 0 }; // we don't care about what gets sent back to us in write mode
+    uint8_t tx[2] = { static_cast<uint8_t> (reg | BMX160_WRITE_BIT), val }; // Set first bit to 0
+
+    HAL_GPIO_WritePin(GPIOG, GPIO_PIN_10, GPIO_PIN_RESET);  // assert slave select
```

Comment:
Do we need to change the value of this GPIO pin after sending and receiving our data?

Reply:
ya, it's the slave select for the IMU chip. Tried getting the SPI hardware to do it, but it kinda deasserts it at weird times and messes up the transmission. People on the internet recommended to just do it in software like this. 

If it's unclear, I could wrap it inside a litte helper function: Assert_SlaveSelect(). DeAssert_SlaveSelect() ?

Reply:
nah all good. Comment explains it :))